text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,169 @@
+
+## Motivation
+
+1. GC worker is an important component for TiKV that deletes outdated MVCC data so as to not explode the storage. But currently, the GC worker is implemented in TiDB, which makes TiKV not usable without TiDB.
+
+2. Flashback is a feature under design for TiDB which allows querying SQL using a specified timestamp. By specifying GC deadline for different tables we can enlarge the use case that allowing arbitrary early timestamp (currently 10 mins) in flashback as long as a proper GC deadline is set for the table.
+
+## Background
+
+According to the [documentation](https://docs.pingcap.com/tidb/stable/garbage-collection-overview) for the **current** GC worker, the GC worker processes the following steps in a GC cycle:
+
+1. Calculate the minimal timestamp among all living SQL sessions, earlier than which are safe to GC, called `Session Safepoint`. Then calculate the `Service Safepoint` for id `gc_worker` by `min(Session Safepoint, now - gc_life_time)`.
+2. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L73) the `gc_worker Service Safepoint` to PD. At the meanwhile, other tools (e.g. CDC, BR) has also pushed their `Service Safepoint` for minimal snapshot timestamp requirement.
+3. Get the minimal `Service Safepoint` among all services from the response of step 2, which is `GC Safepoint`.
+4. Resolve all locks whose timestamp is earlier than the `GC Safepoint`.
+5. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L71) the `GC Safepoint` to PD.
+
+After the `GCSafepoint` being uploaded, TiKV will pull it from PD by interval, then TiKV will automatically delete all records earlier than `GCSafepoint`.
+
+Yet another tricky part is that TiDB has a shortcut to GC a range of keys after `DROP TABLE` or `DROP INDEX`, etc: parallel with step 3 and step 4, TiDB will directly invoke `UnsafeDeleteRange` to TiKV which will delete the range of key-values regardless of locks or timestamp. To make it safe, TiDB will guarantee that the deleted range will never be used again.
+
+## Detailed design
+
+Four new concepts will be introduced by this proposal:
+
+1. `Safeline`
+
+    Similar to `Safepoint`, `Safeline` tells after which timestamp is safe to read and before which is safe to GC, but it will no longer be globally equal among the whole key range but instead separated by key ranges so as to differently push safepoint forward for different key ranges. In this case, safepoint will be a special cased `Safepoint` who ranges across the whole key space.
+
+2. `GC Barrier`
+
+    PD guarantees that using a timestamp later than the `GC Barrier` to read and write on the range within the key range of `GC Barrier` will always be valid. TiKV users (e.g. TiDB / CDC / Client) can put GC Barriers into PD so as to make sure the data they are using will not be GC.
+
+    ![GC Barrier](../media/gc-barrier.png)
```

Comment:
To ensure consistency, in the example in this pic, ResolveLocks should be done on all range above the `safeline`, even in the green area, otherwise it's possible that some unresolved locks' primaries' state is lost after GC. Therefore it would be hard to allow writes in the green area above the `safeline`. So I suggest r/w can be done below the `safeline`, and the green area above `safeline` is readonly.